src/main/scala/org/broadinstitute/dsde/agora/server/business/AgoraBusiness.scala

```diff
@@ -1,126 +1,188 @@
+    if(checkAdmin) {
+      permClient.isAdmin(username) map {
+        case true => permLevel
+        case false => AgoraPermissions(Nothing)
+      }
+    } else {
+      DBIO.successful(AgoraPermissions(Nothing))
     }
+  }
 
-    validatePayload(agoraEntity, username)
+  private def checkNamespacePermission[T](db: DataAccess, agoraEntity: AgoraEntity, username: String, permLevel: AgoraPermissions, checkAdmin: Boolean = false)(op: => ReadWriteAction[T]): ReadWriteAction[T] = {
+    //if we're supposed to check if the user is an admin, create a fake action
+    val admAction = makeDummyAdminPermission(checkAdmin, db.nsPerms, username, permLevel)
+
+    DBIO.sequence(Seq(db.nsPerms.getNamespacePermission(agoraEntity, username), db.nsPerms.getNamespacePermission(agoraEntity, "public"), admAction)) flatMap { namespacePerms =>
+      db.nsPerms.isAdmin(username)
+      if (!namespacePerms.exists(_.hasPermission(permLevel))) {
+        DBIO.failed(NamespaceAuthorizationException(permLevel, agoraEntity, username))
+      } else {
+        op
+      }
+    }
+  }
 
-    val entityToInsert = agoraEntity.entityType.get match {
-      case AgoraEntityType.Configuration =>
-        val method = resolveMethodRef(agoraEntity.payload.get)
-        if (!AgoraEntityPermissionsClient.getEntityPermission(method, username).canRead &&
-            !AgoraEntityPermissionsClient.getEntityPermission(method, "public").canRead) {
-          throw new AgoraEntityNotFoundException(method)
-        }
+  private def checkEntityPermission[T](db: DataAccess, agoraEntity: AgoraEntity, username: String, permLevel: AgoraPermissions)(op: => ReadWriteAction[T]): ReadWriteAction[T] = {
+    DBIO.sequence(Seq(db.aePerms.getEntityPermission(agoraEntity, username), db.aePerms.getEntityPermission(agoraEntity, "public"))) flatMap { entityPerms =>
+      if (!entityPerms.exists(_.hasPermission(permLevel))) {
+        DBIO.failed(AgoraEntityNotFoundException(agoraEntity))
+      } else {
+        op
+      }
+    }
+  }
 
-        agoraEntity.addMethodId(method.id.get.toHexString)
-      case _ => agoraEntity
+  private def createNamespaceIfNonexistent(db: DataAccess, entity: AgoraEntity, entityWithId: AgoraEntity, userAccess: AccessControl): ReadWriteAction[Unit] = {
+    db.nsPerms.doesEntityExists(entity) flatMap { exists =>
+      if(!exists) {
+        db.nsPerms.addEntity(entity) flatMap { _ =>
+          db.nsPerms.insertNamespacePermission(entityWithId, userAccess) map { _ => () }
+        }
+      } else {
+        DBIO.successful(())
+      }
     }
+  }
 
-    val entityWithId = AgoraDao.createAgoraDao(entityToInsert.entityType).insert(entityToInsert.addDate())
-    val userAccess = new AccessControl(username, AgoraPermissions(All))
+  private def checkValidPayload[T](agoraEntity: AgoraEntity, username: String)(op: => ReadWriteAction[T]): ReadWriteAction[T] = {
```

Comment:
used to be [further down](https://github.com/broadinstitute/agora/pull/182/files#diff-1768f9d51119592601a48548da8f9904L96) before i turned it into a check function